buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -313,110 +307,278 @@ void reallocate(int size, int maxCapacity, AdaptiveByteBuf into) {
     }
 
     long usedMemory() {
-        long sum = 0;
-        for (Chunk chunk : centralQueue) {
-            sum += chunk.capacity();
-        }
-        for (Magazine magazine : magazines) {
-            sum += magazine.usedMemory.get();
+        long sum = largeBufferMagazineGroup.usedMemory();
+        for (MagazineGroup group : sizeClassedMagazineGroups) {
+            sum += group.usedMemory();
         }
-        if (liveCachedMagazines != null) {
-            for (Magazine magazine : liveCachedMagazines) {
-                sum += magazine.usedMemory.get();
+        for (MagazineGroup[] groups : allThreadLocalGroups) {
+            for (MagazineGroup group : groups) {
+                sum += group.usedMemory();
             }
         }
         return sum;
     }
 
-    private boolean tryExpandMagazines(int currentLength) {
-        if (currentLength >= MAX_STRIPES) {
-            return true;
+    // Ensure that we release all previous pooled resources when this object is finalized. This is needed as otherwise
+    // we might end up with leaks. While these leaks are usually harmless in reality it would still at least be
+    // very confusing for users.
+    @SuppressWarnings({"FinalizeDeclaration", "deprecation"})
+    @Override
+    protected void finalize() throws Throwable {
```

Comment:
Do we have the cleaners somehow accessible too? 
Maybe they can fit better this

Reply:
When do we need this at all? Only for testing? In that case, maybe just make the allocator Closeable

Reply:
Agree, at that point let users pay their sins :smiling_imp:

Reply:
Another thing that isn't cleaned up is the FastThreadLocal

Reply:
The `finalize` method I've just preserved from the existing code. Whether to keep it or toss it is beyond the scope of this PR. 🫣